File: python/src/onchaindiligence/agent_evidence/dsse.py
"""DSSE v1 Ed25519 sealing helpers."""

from __future__ import annotations

import base64
from collections.abc import Mapping, Sequence
from typing import Any

from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey

from .canonical import canonicalize, parse_json
from .constants import BUNDLE_PAYLOAD_TYPE, BUNDLE_VERSION, MEDIA_TYPE
from .errors import SchemaValidationError, SigningError
from .graph import validate_bundle_payload
from .models import JsonObject
from .schema import validate_document
from .trust import derive_key_id


def dsse_pae(payload_type: str, payload: bytes) -> bytes:
    """Construct DSSE v1 Pre-Authentication Encoding bytes."""

    type_bytes = payload_type.encode("utf-8")
    return b"".join(
        (
            b"DSSEv1 ",
            str(len(type_bytes)).encode("ascii"),
            b" ",
            type_bytes,
            b" ",
            str(len(payload)).encode("ascii"),
            b" ",
            payload,
        )
    )


def load_private_key_pem(value: str | bytes, password: bytes | None = None) -> Ed25519PrivateKey:
    """Load a PKCS8 Ed25519 private key without reading files or configuration."""

    raw = value.encode("utf-8") if isinstance(value, str) else value
    try:
        key = serialization.load_pem_private_key(raw, password=password)
    except (TypeError, ValueError) as exc:
        raise SigningError("private key is not a valid PKCS8 PEM") from exc
    if not isinstance(key, Ed25519PrivateKey):
        raise SigningError("private key must be Ed25519")
    return key


def seal_bundle(
    payload: Mapping[str, Any],
    private_key: Ed25519PrivateKey,
    *,
    key_id: str | None = None,
    keys: Sequence[Mapping[str, Any]] = (),
    registry_snapshots: Sequence[Mapping[str, Any]] = (),
    anchors: Sequence[Mapping[str, Any]] = (),
) -> JsonObject:
    """Seal a valid payload into the portable v0 file without network access."""

    authoritative = dict(payload)
    validate_bundle_payload(authoritative)
    expected_key_id = derive_key_id(private_key.public_key())
    if key_id is not None and key_id != expected_key_id:
        raise SigningError(f"key_id does not match private key; expected {expected_key_id}")
    payload_bytes = canonicalize(authoritative)
    signature = private_key.sign(dsse_pae(BUNDLE_PAYLOAD_TYPE, payload_bytes))
    portable: JsonObject = {
        "media_type": MEDIA_TYPE,
        "bundle_version": BUNDLE_VERSION,
        "envelope": {
            "payloadType": BUNDLE_PAYLOAD_TYPE,
            "payload": base64.b64encode(payload_bytes).decode("ascii"),
            "signatures": [
                {
                    "keyid": expected_key_id,
                    "sig": base64.b64encode(signature).decode("ascii"),
                }
            ],
        },
        "verification_material": {
            "keys": parse_json(canonicalize(list(keys))),
            "registry_snapshots": parse_json(canonicalize(list(registry_snapshots))),
            "anchors": parse_json(canonicalize(list(anchors))),
        },
    }
    try:
        validate_document("portable-file.schema.json", portable)
    except SchemaValidationError as exc:
        raise SigningError(str(exc)) from exc
    return portable

File: packages/agent-evidence/src/dsse.ts
import {
  createPrivateKey,
  createPublicKey,
  KeyObject,
  sign as ed25519Sign,
  verify as ed25519Verify,
} from 'node:crypto'
import { canonicalize, cloneJson } from './canonical.js'
import { BUNDLE_PAYLOAD_TYPE, BUNDLE_VERSION, MEDIA_TYPE } from './constants.js'
import { SchemaValidationError, SigningError } from './errors.js'
import { validateBundlePayload } from './graph.js'
import { validateDocument } from './schema.js'
import { deriveKeyId, loadPublicKey } from './trust.js'
import type {
  AttestationKeyRecord,
  BundlePayload,
  Ed25519Signer,
  JsonObject,
  JsonValue,
  KeyInput,
  PortableBundle,
} from './types.js'

export function dssePae(payloadType: string, payload: Uint8Array): Uint8Array {
  const typeBytes = Buffer.from(payloadType, 'utf8')
  return Buffer.concat([
    Buffer.from('DSSEv1 ', 'ascii'),
    Buffer.from(String(typeBytes.length), 'ascii'),
    Buffer.from(' ', 'ascii'),
    typeBytes,
    Buffer.from(' ', 'ascii'),
    Buffer.from(String(payload.byteLength), 'ascii'),
    Buffer.from(' ', 'ascii'),
    Buffer.from(payload),
  ])
}

function loadPrivateKey(value: KeyInput): KeyObject {
  try {
    const key = value instanceof KeyObject
      ? value
      : createPrivateKey(value instanceof Uint8Array && !Buffer.isBuffer(value) ? Buffer.from(value) : value as string | Buffer)
    if (key.type !== 'private' || key.asymmetricKeyType !== 'ed25519') {
      throw new SigningError('private key must be Ed25519')
    }
    return key
  } catch (error) {
    if (error instanceof SigningError) throw error
    throw new SigningError('private key is not a valid Ed25519 PKCS8 key', { cause: error })
  }
}

export function createEd25519Signer(privateKeyInput: KeyInput): Ed25519Signer {
  const privateKey = loadPrivateKey(privateKeyInput)
  const publicKey = createPublicKey(privateKey)
  return Object.freeze({
    keyId: deriveKeyId(publicKey),
    publicKey,
    sign(message: Uint8Array): Uint8Array {
      return ed25519Sign(null, Buffer.from(message), privateKey)
    },
  })
}

export interface SealBundleOptions {
  keys?: readonly AttestationKeyRecord[]
  registrySnapshots?: readonly JsonObject[]
  anchors?: readonly JsonObject[]
}

export async function sealBundle(
  payload: BundlePayload,
  signer: Ed25519Signer,
  options: SealBundleOptions = {},
): Promise<PortableBundle> {
  validateBundlePayload(payload)
  const publicKey = loadPublicKey(signer.publicKey)
  const expectedKeyId = deriveKeyId(publicKey)
  if (signer.keyId !== expectedKeyId) {
    throw new SigningError(`signer keyId does not match public key; expected ${expectedKeyId}`)
  }
  const payloadBytes = canonicalize(payload)
  const message = dssePae(BUNDLE_PAYLOAD_TYPE, payloadBytes)
  const signature = Buffer.from(await signer.sign(message))
  if (signature.length !== 64 || !ed25519Verify(null, Buffer.from(message), publicKey, signature)) {
    throw new SigningError('signer did not return a valid Ed25519 signature for the DSSE PAE bytes')
  }
  const portable: PortableBundle = {
    media_type: MEDIA_TYPE,
    bundle_version: BUNDLE_VERSION,
    envelope: {
      payloadType: BUNDLE_PAYLOAD_TYPE,
      payload: Buffer.from(payloadBytes).toString('base64'),
      signatures: [{ keyid: signer.keyId, sig: signature.toString('base64') }],
    },
    verification_material: {
      keys: cloneJson([...(options.keys ?? [])] as JsonValue[]) as AttestationKeyRecord[],
      registry_snapshots: cloneJson([...(options.registrySnapshots ?? [])] as JsonValue[]) as JsonObject[],
      anchors: cloneJson([...(options.anchors ?? [])] as JsonValue[]) as JsonObject[],
    },
  }
  try {
    validateDocument('portable-file.schema.json', portable)
  } catch (error) {
    if (error instanceof SchemaValidationError) throw new SigningError(error.message, { cause: error })
    throw error
  }
  return portable
}
